helix-core/src/main/java/org/apache/helix/controller/rebalancer/AbstractRebalancer.java

```diff
@@ -417,4 +417,25 @@ public int compare(String ins1, String ins2) {
       return p1.compareTo(p2);
     }
   }
+
+  // This is for a backward compatible workaround to fix
+  // https://github.com/apache/helix/issues/940.
+  // TODO: remove the workaround once we are able to apply the simple fix without majorly
+  // TODO: impacting user's clusters.
+  protected List<String> getStablePartitionList(ResourceControllerDataProvider clusterData,
+      IdealState currentIdealState) {
+    List<String> partitions =
+        clusterData.getStablePartitionList(currentIdealState.getResourceName());
+    Set<String> currentPartitionSet = currentIdealState.getPartitionSet();
+    if (partitions.size() != currentPartitionSet.size() || !currentPartitionSet
+        .containsAll(partitions)) {
```

Comment:
Let us think about if we  have any other way to do this.  We actually found containsAll() has been a big performance issue before, specially if you compare a set with a list (which take O(N) to iterator).

Reply:
I thought about this, but the other ways are not safe or creating too many additional dependencies. For example, if IS has not been changed, then we don't update the partition lists.
But there is a big assumption there. Doing that might cause some bugs in the future.

I don't have a better idea now. Please suggest.

Reply:
Run a quick profile about containsAll() of Set.

```
  public static void main(String[] args) {
    Set<String> set;
    List<String> input = new ArrayList<>();
    Random ran = new Random();
    for (int i = 0; i < 10000000; i++)  {
      input.add("test" + ran.nextInt() + i);
    }
    long start = System.currentTimeMillis();
    set = new HashSet<>(input);
    long mid = System.currentTimeMillis();
    set.containsAll(input);
    long end = System.currentTimeMillis();
    System.out.println("Phase I: " + (mid - start) + " Phase II: " + (end - mid));
  }
```
The result is Phase I: 5560 Phase II: 1242

So containsAll is not a major concern here. I think.